CoreFeatures/FeatureManagement/UnzipFeatureActor/src/main/java/info/smart_tools/smartactors/feature_management/unzip_feature_actor/UnzipFeatureActor.java

```diff
@@ -51,47 +88,140 @@ public void unzip(final UnzipFeatureWrapper wrapper)
+        if (f.getPath().lastIndexOf(File.separator) >= 0) {
+            location = f.getPath().substring(0, f.getPath().lastIndexOf(File.separator));
         }
         try {
-            ZipFile zipFile = new ZipFile(f);
-            zipFile.extractAll(destination);
-            List<FileHeader> headers = zipFile.getFileHeaders();
-            FileHeader configFileHeader = headers.stream().filter(fh -> fh.getFileName().endsWith("config.json")).findFirst().get();
+            java.nio.file.Path destination = Paths.get(
+                    location, f.getName().substring(0, f.getName().lastIndexOf(EXTENSION_SEPARATOR))
+            );
+            if (checkFeatureWasUpdated(f, feature)) {
+                java.nio.file.Path dir = Files.createDirectories(
+                        destination
+                );
+                Files.copy(
+                        Paths.get(f.getPath()),
+                        Paths.get(dir.toString(), f.getName()),
+                        StandardCopyOption.REPLACE_EXISTING
+                );
+                ZipFile zipFile = new ZipFile(f);
+                List<FileHeader> headers = zipFile.getFileHeaders();
+                FileHeader configFileHeader = headers
+                        .stream()
+                        .filter(fh -> fh.getFileName().endsWith(CONFIG_FILE_NAME))
+                        .findFirst()
+                        .get();
+                if (null != configFileHeader) {
+                    zipFile.extractFile(configFileHeader, dir.toString());
+                }
+            }
 
-            return new File(destination + File.separator + configFileHeader.getFileName());
+            return new File(Paths.get(destination.toString(), CONFIG_FILE_NAME).toString());
         } catch (ZipException e) {
             throw new Exception("Unsupported feature format: broken archive.", e);
         } catch (NoSuchElementException e) {
             throw new Exception("Unsupported feature format: config.json not found or it's broken.", e);
         }
     }
 
+    private boolean checkFeatureWasUpdated(final File zippedFeature, final IFeature feature) {
+        try {
+            File destinationDir = this.getFeatureDirectory(zippedFeature).toFile();
+            if (destinationDir.exists() && destinationDir.isDirectory()) {
+                Long archiveModificationDate = zippedFeature.lastModified();
+                Long newlyFileInUnzippedFeature = Files
+                        .walk(Paths.get(destinationDir.toString()))
+                        .filter(Files::isRegularFile)
+                        .mapToLong(p -> p.toFile().lastModified())
+                        .max().getAsLong();
+                if (archiveModificationDate > newlyFileInUnzippedFeature) {
+                    System.out.println("[INFO] -------------- Unzipping/copying of the feature '" + feature.getName() + "' was skipped because the directory contains an actual state of this feature.");
```

Comment:
What happens if a server process gets interrupted while unzipping a feature (when some files are already unzipped but some are not) and then started again?

Reply:
I think a feature will be loaded not fully. I wanted to add a checksum early but this case was designated as unimportant. I have an idea how this feature should be rewritten and I'll add a checksum in a next refactoring iteration.

Reply:
I have written about another solution to this problem in one of letters earlier.

The problem may be solved the following way:
* when archive is unzipped completely -- create (or modify) a file with a known (e.g. `".unziptimestamp"`) name in directory containing extracted files
* before unzipping -- check modification time of that file and modification time of archive

When server gets interrupted while unzipping a feature the file will not be created/updated.

Resulting code will become more simple (as you will not need to check mtime of all files in a directory) and more reliable. Of course this approach will not fix the issue when extracted files get broken after extracted (e.g. because of broken HDD) but one of motivations for these changes was to provide for developer an ability to change feature files manually what will have the same effect as broken files. So probably we need a strategy that will decide if a feature archive should be extracted and atleast two implementations of it.

Reply:
Ok, I'll fix it.